### PR TITLE
feat: Authentication

### PR DESCRIPTION
**Pull request by vishal804**
Hello,
Created Signin and Signup Functionality
Please review 

Thank You 

### Files to review 

src/PrivateRoute.js
src/component/header/Header.js
src/context/auth-context.js
src/pages/authenticationPage/Signin.js
src/pages/authenticationPage/Signup.js
src/reducer/authReducer.js

[Preview](https://deploy-preview-6--shopping-corner.netlify.app/)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -13,7 +13,6 @@
 
 # misc
 .DS_Store
-.env
```
Comment (ApurvaSawant11) [inline .gitignore]: Secret keys should not be pushed on Github. While deploying the site, you can add this directly over there in environment variables.

File: src/PrivateRoute.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import { useAuth } from "./context/auth-context";
+
+const PrivateRoutes = () => {
+  const { authState } = useAuth();
+  const location = useLocation();
+  if (authState.user) {
+    return <Outlet />;
+  } else {
+    return <Navigate to="/signin" state={{ from: location }} replace />;
+  }
```
Comment (ApurvaSawant11) [inline src/PrivateRoute.js]: Instead of if-else, you can use the ternary operator

File: src/component/categoryCard/CategoryCard.js [hunk 3]
```diff
@@ -8,17 +8,17 @@ const CategoryCard = ({ value }) => {
   const { dispatch } = useData();
   return (
     <>
```
Comment (ApurvaSawant11) [inline src/component/categoryCard/CategoryCard.js]: The extra fragment is not needed here. Everything is already enclosed in the div.
Also check this for all the pages

File: src/component/header/Header.js [hunk 4]
```diff
@@ -1,10 +1,28 @@
...
+  
+
+  const logoutHandler = () => {
+    navigate("/");
+    localStorage.removeItem("userAuthToken");
+    localStorage.removeItem("user");
+    authDispatch({ type: "LOGOUT" });
+  };
+
+  const loginBttonHandler = () => {
```
Comment (ApurvaSawant11) [inline src/component/header/Header.js]: For logout, you are using `logoutHandler`. For login also use the same naming convention. If not, then use the complete `button` spelling

**Review by ApurvaSawant11 — APPROVED** [review]
Looks good to me! Just gave a few suggestions. Check them out.

For UI: The form buttons look bigger than the input field. Have similar padding for both of them. I think it will look better

File: .gitignore (hunk 1 above)
Comment (vishal804) [reply .gitignore]: Done the changes

File: src/PrivateRoute.js (hunk 2 above)
Comment (vishal804) [reply src/PrivateRoute.js]: Done the changes

File: src/component/categoryCard/CategoryCard.js (hunk 3 above)
Comment (vishal804) [reply src/component/categoryCard/CategoryCard.js]: Done the changes

File: src/component/header/Header.js (hunk 4 above)
Comment (vishal804) [reply src/component/header/Header.js]: Done the changes

**Comment by vishal804** [comment]
> Looks good to me! Just gave a few suggestions. Check them out.
> 
> For UI: The form buttons look bigger than the input field. Have similar padding for both of them. I think it will look better

Thanks for Review

File: src/component/card/ProductCard.js [hunk 5]
```diff
@@ -1,13 +1,18 @@
 import React from "react";
-import { useData } from "../../context/data-context";
+import { addToCart, addToWishlist, removeFromWishlist } from "../../utility";
 import { useNavigate } from "react-router-dom";
+import { useData } from "../../context/data-context";
+import { useAuth } from "../../context/auth-context";
```
Comment (nayyyhaa) [inline src/component/card/ProductCard.js:5]: Suggestion: you can import from the absolute path for more clean code. Read: https://dev.to/emanuelone/how-to-use-absolute-path-in-reactjs-503i

File: src/pages/wishlistPage/WishlistPage.js [hunk 6]
```diff
@@ -23,31 +28,35 @@ const WishlistPage = () => {
...
+                <div className="card" key={_id}>
+                  <div className="card-image-container">
+                    <img className="image-responsive" src={Image} alt={alt} />
                   </div>
                   <div>
-                    <div class="card-text-container">
+                    <div className="card-text-container">
                       <p>{brand}</p>
-                      <h5 class="card-heading">{title}</h5>
+                      <h5 className="card-heading">{title}</h5>
```
Comment (nayyyhaa) [inline src/pages/wishlistPage/WishlistPage.js:38]: Appreciation: good that you've changed this to className here.

**Review by nayyyhaa — APPROVED** [review]
Code looks well formated & readable. Found nothing much to suggest. You can merge the PR. 👍

File: src/component/card/ProductCard.js (hunk 5 above)
Comment (vishal804) [reply src/component/card/ProductCard.js:5]: Thanks for suggestion will try to implement that

File: src/pages/wishlistPage/WishlistPage.js (hunk 6 above)
Comment (vishal804) [reply src/pages/wishlistPage/WishlistPage.js:38]: Thank you

**Comment by vishal804** [comment]
> Code looks well formated & readable. Found nothing much to suggest. You can merge the PR. +1

Thanks For Review

**Pull request merged**